digital_asset_types/src/dao/scopes/asset.rs

```diff
@@ -748,6 +753,92 @@ pub async fn get_token_supply(
     })
 }
 
+pub async fn get_token_accounts_by_owner(
+    conn: &impl ConnectionTrait,
+    owner_address: Vec<u8>,
+    mint_address: Option<Vec<u8>>,
+    token_program: Option<Vec<u8>>,
+) -> Result<SolanaRpcResponseAndContext<Vec<RpcData<RpcTokenInfo>>>, DbErr> {
+    let token_accounts = token_accounts::Entity::find();
+
+    let mut conditions = Condition::all().add(token_accounts::Column::Owner.eq(owner_address));
+
+    if let Some(ref mint) = mint_address {
+        conditions = conditions.add(token_accounts::Column::Mint.eq(mint.clone()));
+    }
+
+    if let Some(token_program) = token_program {
+        conditions = conditions.add(token_accounts::Column::TokenProgram.eq(token_program));
+    }
+
+    let token_accounts = token_accounts.filter(conditions).all(conn).await?;
+
+    let mints = if let Some(mint) = mint_address {
+        vec![mint]
+    } else {
+        token_accounts
+            .iter()
+            .map(|ta| ta.mint.clone())
+            .collect::<Vec<_>>()
+    };
+
+    let mint_accounts = tokens::Entity::find()
+        .filter(tokens::Column::Mint.is_in(mints))
+        .all(conn)
+        .await?;
+
+    let mut token_accounts_with_decimals = Vec::new();
+
+    for ta in &token_accounts {
+        let mint_acc_index = mint_accounts.iter().position(|m| m.mint == ta.mint);
+
+        if let Some(mint_acc_index) = mint_acc_index {
+            token_accounts_with_decimals.push((ta.clone(), mint_accounts[mint_acc_index].decimals));
+        }
```

Comment:
Should create a copy that is a hash so iterate once and don't do index scan every loop.